File: backend/app/api/routers/blog.py
"""API ADMIN do blog (P5 §6.B1) — CRUD/publicar pelo studio.

Autenticada (`blog.editar`) e separada da pública (`/api/public/blog`). Aqui é a
mesa de edição do Pablo: criar/editar rascunho, publicar (checkpoint humano),
arquivar. O agente de IA (B2+) escreve via estes mesmos services.
"""
from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile

from app.api.dependencies.auth import require_permission
from app.api.schemas.blog import (
    BlogBriefRequest,
    BlogPautaGerarRequest,
    BlogPautaItem,
    BlogPautaManualCreate,
    BlogPautaUpdate,
    BlogPostAdmin,
    BlogPostCreate,
    BlogPostUpdate,
    BlogRedacao,
    BlogStatusUpdate,
    ChecklistSeoRequest,
    ChecklistSeoResponse,
    GerarImagemRequest,
)
from app.api.services import blog_service
from app.api.services.blog_service import BlogError

router = APIRouter(
    prefix="/api/blog",
    tags=["blog"],
    dependencies=[Depends(require_permission("blog.editar"))],
)


def _handle(e: Exception) -> HTTPException:
    if isinstance(e, BlogError):
        msg = str(e)
        status = 404 if "não encontrad" in msg.lower() else 400
        return HTTPException(status_code=status, detail=msg)
    return HTTPException(status_code=500, detail=f"{type(e).__name__}: {e}")


@router.get("/posts", response_model=list[BlogPostAdmin], summary="Lista posts (todos os status)")
async def listar(status: str | None = Query(None)) -> list[BlogPostAdmin]:
    try:
        return await blog_service.admin.listar(status=status)
    except Exception as e:
        raise _handle(e)


@router.get("/posts/{post_id}", response_model=BlogPostAdmin, summary="Detalhe do post")
async def get(post_id: str) -> BlogPostAdmin:
    try:
        return await blog_service.admin.get(post_id)
    except Exception as e:
        raise _handle(e)


@router.post("/redigir", response_model=BlogRedacao, summary="IA: brief → artigo (rascunho pro editor)")
async def redigir(brief: BlogBriefRequest) -> BlogRedacao:
    try:
        return await blog_service.agente.redigir(brief)
    except Exception as e:
        raise _handle(e)


@router.post("/checklist", response_model=ChecklistSeoResponse, summary="Score SEO do conteúdo (determinístico)")
async def checklist(payload: ChecklistSeoRequest) -> ChecklistSeoResponse:
    try:
        return blog_service.agente.checklist(payload)
    except Exception as e:
        raise _handle(e)


# ── Motor de pauta (B3) ──────────────────────────────────────────
@router.get("/pautas", response_model=list[BlogPautaItem], summary="Backlog de pautas (por score)")
async def listar_pautas(status: str | None = Query(None)) -> list[BlogPautaItem]:
    try:
        return await blog_service.pauta.listar(status)
    except Exception as e:
        raise _handle(e)


@router.post("/pautas/gerar", response_model=list[BlogPautaItem], summary="IA: gera pautas (3 fontes)")
async def gerar_pautas(req: BlogPautaGerarRequest) -> list[BlogPautaItem]:
    try:
        return await blog_service.pauta.gerar(req)
    except Exception as e:
        raise _handle(e)


@router.post("/pautas/{pauta_id}/escrever", response_model=BlogPostAdmin, summary="1 clique: pauta → rascunho (IA)")
async def escrever_pauta(pauta_id: str) -> BlogPostAdmin:
    try:
        return await blog_service.coordenador.escrever_pauta(pauta_id)
    except Exception as e:
        raise _handle(e)


@router.post("/pautas", response_model=BlogPautaItem, status_code=201, summary="Cria pauta manual")
async def criar_pauta(payload: BlogPautaManualCreate) -> BlogPautaItem:
    try:
        return await blog_service.pauta.criar_manual(payload)
    except Exception as e:
        raise _handle(e)


@router.put("/pautas/{pauta_id}", response_model=BlogPautaItem, summary="Edita pauta (status/score/…)")
async def atualizar_pauta(pauta_id: str, payload: BlogPautaUpdate) -> BlogPautaItem:
    try:
        return await blog_service.pauta.atualizar(pauta_id, payload)
    except Exception as e:
        raise _handle(e)


@router.delete("/pautas/{pauta_id}", status_code=204, summary="Apaga pauta")
async def deletar_pauta(pauta_id: str) -> None:
    try:
        await blog_service.pauta.deletar(pauta_id)
    except Exception as e:
        raise _handle(e)


@router.post("/posts", response_model=BlogPostAdmin, status_code=201, summary="Cria post (rascunho)")
async def criar(payload: BlogPostCreate) -> BlogPostAdmin:
    try:
        return await blog_service.admin.criar(payload)
    except Exception as e:
        raise _handle(e)


@router.put("/posts/{post_id}", response_model=BlogPostAdmin, summary="Edita post (patch parcial)")
async def atualizar(post_id: str, payload: BlogPostUpdate) -> BlogPostAdmin:
    try:
        return await blog_service.admin.atualizar(post_id, payload)
    except Exception as e:
        raise _handle(e)


@router.post("/posts/{post_id}/imagem", response_model=BlogPostAdmin, summary="IA: gera imagem (capa/seção) → MinIO")
async def gerar_imagem(post_id: str, req: GerarImagemRequest) -> BlogPostAdmin:
    try:
        return await blog_service.imagens.gerar(
            post_id, papel=req.papel, prompt=req.prompt, aspect_ratio=req.aspect_ratio
        )
    except Exception as e:
        raise _handle(e)


@router.post("/posts/{post_id}/imagem/upload", response_model=BlogPostAdmin, summary="Sobe a imagem final (editada fora)")
async def upload_imagem(
    post_id: str,
    arquivo: UploadFile = File(...),
    papel: str = Form("cover"),
    alt: str | None = Form(None),
) -> BlogPostAdmin:
    try:
        data = await arquivo.read()
        return await blog_service.imagens.upload(
            post_id,
            papel=papel,
            data=data,
            content_type=arquivo.content_type or "image/png",
            alt=alt,
        )
    except Exception as e:
        raise _handle(e)


@router.patch("/posts/{post_id}/status", response_model=BlogPostAdmin, summary="Muda status (publicar/arquivar)")
async def mudar_status(post_id: str, payload: BlogStatusUpdate) -> BlogPostAdmin:
    try:
        return await blog_service.admin.mudar_status(post_id, payload.status)
    except Exception as e:
        raise _handle(e)


@router.delete("/posts/{post_id}", status_code=204, summary="Apaga post")
async def deletar(post_id: str) -> None:
    try:
        await blog_service.admin.deletar(post_id)
    except Exception as e:
        raise _handle(e)

File: frontend/lib/api/blog.ts
// Agente Blog (P5 §6) — admin no studio: CRUD + publicar. A leitura pública
// (site) usa /api/public/blog; aqui é a mesa de edição autenticada.

import { request } from './client';
import type {
  BlogBriefRequest,
  BlogPauta,
  BlogPautaGerarRequest,
  BlogPautaManualCreate,
  BlogPautaStatus,
  BlogPautaUpdate,
  BlogPostAdmin,
  BlogPostCreate,
  BlogPostUpdate,
  BlogRedacao,
  BlogStatus,
  CapaSugestoesResponse,
  ChecklistSeo,
  ChecklistSeoRequest,
  GerarImagemConteudoRequest,
  GerarImagemRequest,
  ImagemConteudoSugestoesResponse,
} from '../types';

const BASE = '/api/blog';
const T = 12_000;

export const blogApi = {
  blogListar(status?: BlogStatus): Promise<BlogPostAdmin[]> {
    const qs = status ? `?status=${encodeURIComponent(status)}` : '';
    return request<BlogPostAdmin[]>(`${BASE}/posts${qs}`, { timeoutMs: T });
  },

  blogGet(id: string): Promise<BlogPostAdmin> {
    return request<BlogPostAdmin>(`${BASE}/posts/${id}`, { timeoutMs: T });
  },

  blogCriar(body: BlogPostCreate): Promise<BlogPostAdmin> {
    return request<BlogPostAdmin>(`${BASE}/posts`, {
      method: 'POST',
      body,
      timeoutMs: T,
    });
  },

  blogAtualizar(id: string, body: BlogPostUpdate): Promise<BlogPostAdmin> {
    return request<BlogPostAdmin>(`${BASE}/posts/${id}`, {
      method: 'PUT',
      body,
      timeoutMs: T,
    });
  },

  blogMudarStatus(id: string, status: BlogStatus): Promise<BlogPostAdmin> {
    return request<BlogPostAdmin>(`${BASE}/posts/${id}/status`, {
      method: 'PATCH',
      body: { status },
      timeoutMs: T,
    });
  },

  blogRemover(id: string): Promise<void> {
    return request<void>(`${BASE}/posts/${id}`, {
      method: 'DELETE',
      timeoutMs: T,
    });
  },

  // ── Agente (B2) ─────────────────────────────────────────────────
  blogRedigir(brief: BlogBriefRequest): Promise<BlogRedacao> {
    return request<BlogRedacao>(`${BASE}/redigir`, {
      method: 'POST',
      body: brief,
      timeoutMs: 90_000, // geração de artigo pela LLM é lenta
    });
  },

  blogChecklist(payload: ChecklistSeoRequest): Promise<ChecklistSeo> {
    return request<ChecklistSeo>(`${BASE}/checklist`, {
      method: 'POST',
      body: payload,
      timeoutMs: T,
    });
  },

  // ── Pautas (B3) ─────────────────────────────────────────────────
  blogPautas(status?: BlogPautaStatus): Promise<BlogPauta[]> {
    const qs = status ? `?status=${encodeURIComponent(status)}` : '';
    return request<BlogPauta[]>(`${BASE}/pautas${qs}`, { timeoutMs: T });
  },

  blogGerarPautas(req: BlogPautaGerarRequest): Promise<BlogPauta[]> {
    return request<BlogPauta[]>(`${BASE}/pautas/gerar`, {
      method: 'POST',
      body: req,
      timeoutMs: 90_000,
    });
  },

  /** Coordenador 1-clique: pauta → rascunho completo (IA escreve e salva). */
  blogEscreverPauta(id: string): Promise<BlogPostAdmin> {
    return request<BlogPostAdmin>(`${BASE}/pautas/${id}/escrever`, {
      method: 'POST',
      timeoutMs: 90_000,
    });
  },

  // ── Imagens (B-IMG) ─────────────────────────────────────────────
  blogSugerirCapas(id: string): Promise<CapaSugestoesResponse> {
    return request<CapaSugestoesResponse>(`${BASE}/posts/${id}/imagem/sugestoes`, {
      method: 'POST',
      timeoutMs: 60_000,
    });
  },

  blogGerarImagem(id: string, body: GerarImagemRequest): Promise<BlogPostAdmin> {
    return request<BlogPostAdmin>(`${BASE}/posts/${id}/imagem`, {
      method: 'POST',
      body,
      timeoutMs: 120_000,
    });
  },

  /** Gera as imagens do corpo (marcadores {{IMG}}) — pode levar ~1min p/ 2-3. */
  blogGerarImagensConteudo(id: string): Promise<BlogPostAdmin> {
    return request<BlogPostAdmin>(`${BASE}/posts/${id}/imagem/conteudo`, {
      method: 'POST',
      timeoutMs: 180_000,
    });
  },

  /** Sugere imagens pro corpo (por seção) — igual as capas. */
  blogSugerirImagensConteudo(id: string): Promise<ImagemConteudoSugestoesResponse> {
    return request<ImagemConteudoSugestoesResponse>(
      `${BASE}/posts/${id}/imagem/conteudo/sugestoes`,
      { method: 'POST', timeoutMs: 60_000 },
    );
  },

  /** Gera 1 imagem de conteúdo (sugestão escolhida) e insere na seção. */
  blogInserirImagemConteudo(
    id: string,
    body: GerarImagemConteudoRequest,
  ): Promise<BlogPostAdmin> {
    return request<BlogPostAdmin>(`${BASE}/posts/${id}/imagem/conteudo/inserir`, {
      method: 'POST',
      body,
      timeoutMs: 120_000,
    });
  },

  blogUploadImagem(
    id: string,
    arquivo: File,
    papel: 'cover' | 'secao' = 'cover',
    alt?: string,
  ): Promise<BlogPostAdmin> {
    const fd = new FormData();
    fd.append('arquivo', arquivo);
    fd.append('papel', papel);
    if (alt) fd.append('alt', alt);
    return request<BlogPostAdmin>(`${BASE}/posts/${id}/imagem/upload`, {
      method: 'POST',
      body: fd,
      timeoutMs: 60_000,
    });
  },

  blogCriarPauta(body: BlogPautaManualCreate): Promise<BlogPauta> {
    return request<BlogPauta>(`${BASE}/pautas`, {
      method: 'POST',
      body,
      timeoutMs: T,
    });
  },

  blogAtualizarPauta(id: string, body: BlogPautaUpdate): Promise<BlogPauta> {
    return request<BlogPauta>(`${BASE}/pautas/${id}`, {
      method: 'PUT',
      body,
      timeoutMs: T,
    });
  },

  blogRemoverPauta(id: string): Promise<void> {
    return request<void>(`${BASE}/pautas/${id}`, {
      method: 'DELETE',
      timeoutMs: T,
    });
  },
};
